Senlin: Actions List

JackKuei opened the pull request:
Prior to a PR being reviewed, there needs to be a Github issue that the PR
addresses. Replace the brackets and text below with that issue number.

For #[823 [Add support for Senlin clustering]](https://github.com/gophercloud/gophercloud/issues/823)

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

[[actions:list]](https://github.com/openstack/senlin/blob/e5bc16b3fea2f10dead365f983c5a937f23d72b0/senlin/api/openstack/v1/actions.py#L96)

https://github.com/openstack/senlin/blob/98d8cf30a374858079eed653b2acbf9dd5d57318/senlin/objects/action.py#L29

https://github.com/openstack/senlin/blob/98d8cf30a374858079eed653b2acbf9dd5d57318/senlin/objects/action.py#L179

https://github.com/openstack/senlin/blob/bdf4fc024a63b63f43900378a113a808b1c3ef3d/senlin/api/openstack/v1/actions.py#L59-L68

jtopjian commented:
@JackKuei Looks like the Travis failure is due to a `gofmt` issue.

JackKuei commented on openstack/clustering/v1/actions/results.go:
```diff
@@ -0,0 +1,108 @@
+	Name         string                   `json:"name"`
+	Outputs      map[string]interface{}   `json:"outputs"`
+	Owner        string                   `json:"owner"`
+	ProjectUUID  string                   `json:"project"`
```
The reason having "UUID" in the name so the caller doesn't have to know if this is ID, UUID, or name. This also make it clear from the opts, such as CreateOpts where the UserID can be user name/user UUID/user ID

jtopjian replied:
I agree about the user friendliness aspect, but the field names must match the tag names.

Making the options more friendly should be the role of the application that consumes gophercloud.

JackKuei commented:
Sorry missed the first part. Yes, all these have been tested on the real environment using pike.

JackKuei commented on openstack/clustering/v1/actions/results.go:
```diff
@@ -0,0 +1,108 @@
+
+type JSONRFC3339Milli time.Time
+
+const RFC3339Milli = "2006-01-02T15:04:05.999999Z"
```
Just remembered why have to use custom time parsing for JSONRFC3339Milli. Currently if using gophercloud.JSONRFC3339Milli and Senlin returns empty for the field, the parsing would fail. So I added this inside UnmarshalJSON for JSONRFC3339Milliif string(data) == "null" { *jt = JSONRFC3339Milli(time.Time{}) return nil } Should supporting empty for the time field be added into gophercloud.JSONRFC3339Milli or be in the results.go for clustering only?

jtopjian commented:
@JackKuei This sounds similar to what @dkt26111 did here:

https://github.com/gophercloud/gophercloud/blob/master/openstack/clustering/v1/policies/results.go#L68-L102

jtopjian commented:
@JackKuei Also if there are results which cause different code paths to happen, we'll want to include unit tests for those. In this case, we need to make sure that all time formats are tested.

JackKuei commented:
Decided to wrap the gophercloud.JSONRFC3339Milli but allows empty and null json field

JackKuei commented:
Still need to check multi-page. Currently seems everything is within 1 page even when there are 3000+ actions

jtopjian commented:
> Decided to have customized JSONRFC3339Milli which wraps the gophercloud.JSONRFC3339Milli but allow empty and null json field. Also added unit test for these

Thank you for adding the test case.

It's not necessary to create a separate type. This can all be handled in the Action's `UnmarshalJSON` method. If you're seeing both a combination of `null` and `""` values for the `CreatedAt` and `UpdatedAt` fields, then you can handle everything with this:

```go
func (r *Action) UnmarshalJSON(b []byte) error {
  type tmp Action
  var s struct {
    tmp
    CreatedAt interface{} `json:"created_at"`
    UpdatedAt interface{} `json:"updated_at"`
  }
  err := json.Unmarshal(b, &s)
  if err != nil {
    return err
  }
  *r = Action(s.tmp)

  switch t := s.CreatedAt.(type) {
  case string:
    if t != "" {
      r.CreatedAt, err = time.Parse(gophercloud.RFC3339Milli, t)
      if err != nil {
        return err
      }
    }
  }

  switch t := s.UpdatedAt.(type) {
  case string:
    if t != "" {
      r.UpdatedAt, err = time.Parse(gophercloud.RFC3339Milli, t)
      if err != nil {
        return err
      }
    }
  }

  return nil
}
```

The above adds a case for strings which are not `""`. If the string _is_ `""`, then it's not handled at all which implicitly sets the fields to `nil` (or a zero-value time).

There's no need to handle `null` explicitly because that gets implicitly converted to `nil` (or a zero-value time).

The above passes with the test fixture you provided.

Let me know if this isn't working for you, though, or if you have any questions about the implementation.

JackKuei commented:
Thanks for the code snippet. But isn't using the gophercloud UnmarshalJSON a cleaner choice? Because in this case, would also need to add default case to the switch statement to return error if somehow json contains any other types beside string for the time fields. Also if there are three time fields, such as CreateAt, InitAt, UpdateAt then would need to duplicate the time parsing code blocks 3 times, doesn't seem too elegant. Also if the gophercloud's UnmarshalJSON changed to a even better way of parsing, then would get the benefits for free.

jtopjian commented:
The elegant solution would be the Senlin API not sending back multiple formats of a time field. 😉 

In terms of design, we are localizing the workarounds required to handle the API results within the API result struct itself. We don't want to patch Gophercloud to handle inconsistent requests globally.

With regard to multiple blocks of code, Go is not a language where the smallest number of lines is the better solution (the repetitiveness of `if err != nil` is a good example of this). If it takes 10 blocks of code to smooth out 10 fields in a struct, so be it. At least then, the person reading the code is able to see what's happening within that package itself rather than not realizing there's magic going on in another package.

The pattern of handling the API result's quirks within the result struct's `UnmarshalJSON` method is the standard solution for handling this in Gophercloud. If you review other implementations of `UnmarshalJSON` throughout Gophercloud, you'll find various ways of handling different type conversions. There might very well be a better way of handling this time conversion than what I have proposed and I'm certainly open to other solutions. However, I can't budge from doing this outside of the result struct's `UnmarshalJSON` method.

JackKuei commented:
Yes, my two major gripes with senlin is the different types of time fields and the inconsistent naming for ID/UUID/Names. Created an extreme example having 6 time fields for comparisons.

JackKuei commented:
`func (r *Action) UnmarshalJSON(b []byte) error {
	type tmp Action
	var s struct {
		tmp
		CreatedAt interface{} `json:"created_at"`
		UpdatedAt interface{} `json:"updated_at"`
		InitAt    interface{} `json:"updated_at"`
		Test1     interface{} `json:"test1"`
		Test2     interface{} `json:"test2"`
		Test3     interface{} `json:"test3"`
	}
	err := json.Unmarshal(b, &s)
	if err != nil {
		return err
	}
	*r = Action(s.tmp)

	switch t := s.CreatedAt.(type) {
	case string:
		if t != "" {
			r.CreatedAt, err = time.Parse(gophercloud.RFC3339Milli, t)
			if err != nil {
				return err
			}
		}
	case nil:
		return nil
	default:
		return fmt.Errorf("type=%v", reflect.TypeOf(s.CreatedAt))
	}

	switch t := s.UpdatedAt.(type) {
	case string:
		if t != "" {
			r.UpdatedAt, err = time.Parse(gophercloud.RFC3339Milli, t)
			if err != nil {
				return err
			}
		}
	case nil:
		return nil
	default:
		return fmt.Errorf("type=%v", reflect.TypeOf(s.UpdatedAt))
	}

	switch t := s.InitAt.(type) {
	case string:
		if t != "" {
			r.InitAt, err = time.Parse(gophercloud.RFC3339Milli, t)
			if err != nil {
				return err
			}
		}
	case nil:
		return nil
	default:
		return fmt.Errorf("Invalid Type %+v", reflect.TypeOf(s.UpdatedAt))
	}

	switch t := s.Test1.(type) {
	case string:
		if t != "" {
			r.Test1, err = time.Parse(gophercloud.RFC3339Milli, t)
			if err != nil {
				return err
			}
		}
	case nil:
		return nil
	default:
		return fmt.Errorf("Invalid Type %+v", reflect.TypeOf(s.Test1))
	}

	switch t := s.Test2.(type) {
	case string:
		if t != "" {
			r.Test2, err = time.Parse(gophercloud.RFC3339Milli, t)
			if err != nil {
				return err
			}
		}
	case nil:
		return nil
	default:
		return fmt.Errorf("Invalid Type %+v", reflect.TypeOf(s.Test2))
	}

	switch t := s.Test3.(type) {
	case string:
		if t != "" {
			r.Test3, err = time.Parse(gophercloud.RFC3339Milli, t)
			if err != nil {
				return err
			}
		}
	case nil:
		return nil
	default:
		return fmt.Errorf("Invalid Type %+v", reflect.TypeOf(s.Test3))
	}

	return nil
}
`

JackKuei commented:
`type JSONRFC3339MilliAllowEmpty gophercloud.JSONRFC3339Milli

func (jt *JSONRFC3339MilliAllowEmpty) UnmarshalJSON(b []byte) error {
	d := string(b)
	d = strings.Trim(d, "\"' ")
	if len(d) == 0 || d == "null" {
		// Allows empty or "null" field
		*jt = JSONRFC3339MilliAllowEmpty(time.Time{})
		return nil
	}

	gt := gophercloud.JSONRFC3339Milli{}
	//err := gt.UnmarshalJSON(b)
	err := json.Unmarshal(b, &gt)
	if err != nil {
		return err
	}
	*jt = JSONRFC3339MilliAllowEmpty(gt)
	return err
}

func (r *Action) UnmarshalJSON(b []byte) error {
	type tmp Action
	var s struct {
		tmp
		CreatedAt JSONRFC3339MilliAllowEmpty `json:"created_at"`
		UpdatedAt JSONRFC3339MilliAllowEmpty `json:"updated_at"`
		InitAt    JSONRFC3339MilliAllowEmpty `json:"updated_at"`
		Test1     JSONRFC3339MilliAllowEmpty `json:"test1"`
		Test2     JSONRFC3339MilliAllowEmpty `json:"test2"`
		Test3     JSONRFC3339MilliAllowEmpty `json:"test3"`
	}
	err := json.Unmarshal(b, &s)
	if err != nil {
		return err
	}
	*r = Action(s.tmp)

	r.CreatedAt = time.Time(s.CreatedAt)
	r.UpdatedAt = time.Time(s.UpdatedAt)
	r.InitAt = time.Time(s.UpdatedAt)
	r.Test1 = time.Time(s.UpdatedAt)
	r.Test2 = time.Time(s.UpdatedAt)
	r.Test3 = time.Time(s.UpdatedAt)

	return nil
}
`

jtopjian commented:
>  Need to have nil in switch because otherwise will fall into the default case which returns error for such types as float, int etc.

Is it really possible for Senlin to send back time as integers and floats? I can understand having to account for timestamps, strings, and nulls.

JackKuei commented:
It shouldn't return any other time except for time string. ```The date and time when the object was created. The date and time stamp format is ISO8601: CCYY-MM-DDThh:mm:ssZ. For example: 2016-01-18T00:00:00Z``` Although the start_time and end_time fields are in float

jtopjian commented:
> Although the start_time and end_time fields are in float

Ah yes, you're right. It looks like they're in epoch format. 

Is there benefit to having these in epoch/float format? If not, we should also convert these to `time.Time`.

JackKuei commented:
Probably in float to easier calculate time span. I feel it is easier for the gophercloud user to use time.Time rather than float. But Senlin doc specifies float.

jtopjian commented:
> But Senlin doc specifies float.

Yeah, that's a good point... OK, let's keep it float for now. I'm thinking there's intention behind making it epoch and not a timestamp, so let's keep it that way.

jtopjian commented on openstack/clustering/v1/actions/results.go:
```diff
@@ -0,0 +1,116 @@
+			}
+		}
+	case nil:
+		// Valid continue
```
This is technically correct, but let's side on explicitness:

```go
case nil:
  r.CreatedAt = time.Time{}
```

jtopjian commented on openstack/clustering/v1/actions/results.go:
```diff
@@ -0,0 +1,116 @@
+			}
+		}
+	case nil:
+		// Valid continue
```
Same here

animationzl commented:
recheck

JackKuei commented:
- Reverted back to use float for StartTime and EndTime. 
- Reverted accident check-in of:
     openstack/compute/v2/extensions/quotasets/testing/fixtures.go
     openstack/networking/v2/extensions/vpnaas/ikepolicies/testing/requests_test.go 
- Explicitly use time.Time{} when json is null
